core/sr-staking-primitives/src/offence.rs

```diff
@@ -0,0 +1,147 @@
+	/// along with the `session_index`. Two offences considered to happen at the same time iff
+	/// both `session_index` and `time_slot` are equal.
+	///
+	/// As an example, for GRANDPA timescale could be a round number and for BABE it could be a slot
+	/// number. Note that for BABE the round number is reset each epoch.
+	fn time_slot(&self) -> TimeSlot;
+
+	/// A slash fraction of the total exposure that should be slashed for this
+	/// particular offence kind for the given parameters.
+	///
+	/// `offenders_count` - the count of unique offending authorities. It is >0.
+	/// `validator_set_count` - the cardinality of the validator set at the time of offence.
+	fn slash_fraction(
+		offenders_count: u32,
+		validator_set_count: u32,
+	) -> Perbill;
+}
+
+/// A trait for decoupling offence reporters from the actual handling of offence reports.
+pub trait ReportOffence<Reporter, Offender, O: Offence<Offender>> {
+	/// Report an `offence` and reward the `reporter`.
+	fn report_offence(reporter: Option<Reporter>, offence: O);
+}
+
+impl<Reporter, Offender, O: Offence<Offender>> ReportOffence<Reporter, Offender, O> for () {
+	fn report_offence(_reporter: Option<Reporter>, _offence: O) {}
+}
+
+/// A trait to take action on an offence.
+///
+/// Used to decouple the module that handles offences and
+/// the one that should punish for those offences.
+pub trait OnOffenceHandler<Reporter, Offender> {
+	/// A handler for offence of particular kind.
+	///
+	/// Note that this contains a list of all previous offenders
+	/// as well. The implementer should cater for a case, where
+	/// the same authorities were reported for the same offence
+	/// in the past (see `OffenceCount`).
+	///
+	/// The vector of `slash_fraction` contains perbils
+	/// the authorities should be slashed and is computed
+	/// according to the `OffenceCount` already. This is of the same length as `offenders.`
+	/// Zero is a valid value for a fraction.
+	fn on_offence(
+		offenders: &[OffenceDetails<Reporter, Offender>],
+		slash_fraction: &[Perbill],
+	);
+}
+
+impl<Reporter, Offender> OnOffenceHandler<Reporter, Offender> for () {
+	fn on_offence(
+		_offenders: &[OffenceDetails<Reporter, Offender>],
+		_slash_fraction: &[Perbill],
+	) {}
+}
+
+/// A details about an offending authority for a particular kind of offence.
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct OffenceDetails<Reporter, Offender> {
+	/// The offending authority id
+	pub offender: Offender,
+	/// A number of times the authority was already reported for this offence.
+	///
+	/// Since we punish all authorities for that made the same offence in the past as well
+	/// we keep track the "age" of the report, so that the slashes can be lowered
+	/// in case the authority was already slashed in the past.
+	/// Note that we don't buffer slashes and instead use this approach.
+	pub count: OffenceCount,
```

Comment:
is the intended purpose of this member to make it so an authority slashed for amount `f(a)` before, and now who would receive a slash of `f(b)` actually receives a slash of `f(b) - f(a)`?

i.e., not to over-step the slash amount function when someone is slashed multiple times

If so, wouldn't we also need to know how many other authorities had performed the same offence at the time of the prior slash?

Reply:
since offenders are reported one at a time, we know that previous slashes will account for `count - 1` slashes to all previous offenders, no?

Reply:
> A number of times the authority was already reported for this offence.

I understand this documentation line to mean only the `count` of _reports_ for the same authority `Self: OffenceDetails` is implicitly associated with.

If it is actually the `count` of slashes the authority has received (whether as a result of its own misbehavior or others'), then yes.

Reply:
(although, if that is the case, then the `count` parameter will be the same across all offenders in the same time-slot and should probably be refactored elsewhere)